packages/perseus/src/widgets/interactive-graphs/reducer/interactive-graph-reducer.ts

```diff
@@ -8,60 +8,53 @@ import {
     MOVE_CONTROL_POINT,
     MOVE_LINE,
     MOVE_POINT,
+    MoveAll,
+    MoveControlPoint,
+    MoveLine,
+    MovePoint,
 } from "./interactive-graph-action";
 
 import type {CollinearTuple} from "../../../perseus-types";
-import type {InteractiveGraphState} from "../types";
+import type {InteractiveGraphState, PairOfPoints} from "../types";
 import type {Interval} from "mafs";
 
 /** Determine if coords is type CollinearTuple[] */
 const isCollinearTuples = (
     coords: readonly CollinearTuple[] | readonly vec.Vector2[],
 ): coords is readonly CollinearTuple[] => Array.isArray(coords[0][0]);
 
-// Generic type makes returned state match input state
-export function interactiveGraphReducer<
-    GraphState extends InteractiveGraphState,
->(state: Readonly<GraphState>, action: InteractiveGraphAction): GraphState {
+function doMoveControlPoint(
+    state: InteractiveGraphState,
+    action: MoveControlPoint,
+): InteractiveGraphState {
     const {snapStep, range} = state;
-    if (!state.coords) {
-        throw new Error("Graph state must have been initialized with coords");
-    }
-    switch (action.type) {
-        case MOVE_CONTROL_POINT: {
-            const newCoords =
-                action.itemIndex !== undefined &&
-                state.coords &&
-                isCollinearTuples(state.coords)
-                    ? updateAtIndex({
-                          array: state.coords,
-                          index: action.itemIndex,
-                          update: (tuple) =>
-                              setAtIndex({
-                                  array: tuple,
-                                  index: action.pointIndex,
-                                  newValue: snap({
-                                      snapStep,
-                                      point: bound({
-                                          snapStep,
-                                          range,
-                                          point: action.destination,
-                                      }),
-                                  }),
-                              }),
-                      })
-                    : setAtIndex({
-                          array: state.coords,
-                          index: action.pointIndex,
-                          newValue: snap({
-                              snapStep,
-                              point: bound({
-                                  snapStep,
-                                  range,
-                                  point: action.destination,
-                              }),
-                          }),
-                      });
+    switch (state.type) {
+        case "segment":
+        case "linear":
+        case "linear-system":
+        case "ray": {
+            if (action.itemIndex == null) {
+                throw new Error(
+                    "MoveControlPoint.itemIndex cannot be null when moving a point on a line",
```

Comment:
The need for this exception is a sign that the types still aren't quite right. @mark-fitzgerald maybe we can pair on addressing this.